Some tweaks to the Getting Started docs (#2195)
EnzymeAD/Enzyme.jl: docs/src/index.md

We will try a few things with the following functions:

```jldoctest rosenbrock
julia> rosenbrock(x, y) = (1.0 - x)^2 + 100.0 * (y - x^2)^2;

julia> rosenbrock_inp(x) = (1.0 - x[1])^2 + 100.0 * (x[2] - x[1]^2)^2;
```

> where we note for future reference that the value of this function at `x=1.0`, `y=2.0` is `100.0`, and its derivative

Reviewer: Consider showing this as code instead of prose? 
```
julia> rosenbrock(x, y) = (1.0 - x)^2 + 100.0 * (y - x^2)^2;

julia> rosenbrock_inp(xy) = (1.0 - xy[1])^2 + 100.0 * (xy[2] - xy[1]^2)^2;

julia> rosenbrock(1.0, 2.0) == rosenbrock_inp([1.0, 2.0]) == 100.0
true
```
I also think you should not call the input of `rosenbrock_inp` the same thing, `x == [x, y]` is weird. The name `rosenbrock_inp` also seems a bit weird, maybe it can just be another method, or if that's too confusing, add a suffix more informative than "inp"? (I'm not sure what INP means, maybe input, but why?)

Reviewer 2: I think it was originally in place (@michel2323 were you the one to originally author this doc, just by virtue of it being rosenbrock?)

But either way sure!

Reviewer: Ah ok. But this function isn't in-place, it's just going to be used somewhere below in a demonstration that Enzyme likes to handle functions which accept Vector by mutating something else. The reader doesn't know that yet.

Reviewer 2: very true, maybe rosenbrok_array or something? or even just rosenbrock2

> julia> rosenbrock_inp(x) = (1.0 - x[1])^2 + 100.0 * (x[2] - x[1]^2)^2;
> ```
> 
> where we note for future reference that the value of this function at `x=1.0`, `y=2.0` is `100.0`, and its derivative
> with respect to `x` at that point is `-400.0`, and its derivative with respect to `y` at that point is `200.0`. 

Reviewer: ````suggestion
julia> rosenbrock(xy::Vector) = (1.0 - xy[1])^2 + 100.0 * (xy[2] - xy[1]^2)^2;

julia> z = rosenbrock(1.0, 2.0)
100.0

julia> z == rosenbrock([1.0, 2.0])  # Vector method
true
```
We note for future reference that the value of this function at `x=1.0`, `y=2.0` is `z=100.0`. Its derivative with respect to `x` at that point is `-400.0`, and its derivative with respect to `y` is `200.0`. 
````

## Reverse mode

The return value of reverse mode [`autodiff`](@ref) is a tuple that contains as a first value
the derivative value of the active inputs and optionally the _primal_ return value (i.e. the 

> value of the undifferentiated function).

Reviewer: Consider not using "value" to mean so many things here?

> is a tuple that contains as a first *element* the *derivatives* of ..., and optionally the primal *value* (i.e. what the function returns).

Reviewer: Maybe "optionally" also seems a bit odd to describe the output not the input. It's not that you may omit this. It's that ReverseWithPrimal tells it to.

Reviewer 2: yeah we definitely don't need to say "derivative value" and can just say "the derivative of"

Reviewer 2: and "the value of the undifferentiated function" -> "the result of the original function without differentiation"

Reviewer: Also consider putting the ReverseWithPrimal case first, as without it,  `((-400.0, 200.0),)` seems like a puzzle to count the brackets & guess why.

Reviewer 2: oh yeah totally fair, if you want to put that in this PR that would be fine with me!

Reviewer: I can't make suggestions across deleted lines :/  so this is going to be messy...

> the derivative value of the active inputs and optionally the _primal_ return value (i.e. the 
> value of the undifferentiated function).

Reviewer: ```suggestion
the derivatives with respect to the inputs. The tuple's second element is the _primal_ value (i.e. the result of the original function without differentiation),
but this is omitted if you use `Reverse` instead of `ReverseWithPrimal`:
```

```jldoctest rosenbrock

> julia> autodiff(Reverse, rosenbrock, Active(1.0), Active(2.0))

Reviewer: ```suggestion
julia> derivs, z = autodiff(ReverseWithPrimal, rosenbrock, Active(1.0), Active(2.0))
((-400.0, 200.0), 100.0)

julia> autodiff(Reverse, rosenbrock, Active(1.0), Active(2.0))
```

((-400.0, 200.0),)

> julia> autodiff(ReverseWithPrimal, rosenbrock, Active(1.0), Active(2.0))
> ((-400.0, 200.0), 100.0)

Reviewer: ```suggestion
```

 0.0
 0.0

> julia> autodiff(Reverse, rosenbrock_inp, Duplicated(x, dx))

Reviewer: ```suggestion
julia> autodiff(Reverse, rosenbrock, Duplicated(x, dx))
```

  200.0
```

> Both the inplace and "normal" variant return the gradient. The difference is that with [`Active`](@ref) 

Reviewer: This wording seems weird. The inplace version returns `((nothing,),)`, it's written right there. And inplace / "normal" are new terms here, which aren't the terms you need to learn to understand Enzyme.

> The version with `Active` (for immutable inputs like x::Float64) returns the gradient. The version with `Duplicated` (for mutable inputs like `x::Vector{Float64}`) instead writes the gradient into the `Duplicated` object, and returns `nothing` in the corresponding slot of the returned `derivs`.

Reviewer: ```suggestion
Both versions calculate the same derivatives. The difference is that with [`Active`](@ref) arguments
```

> the gradient is returned and with [`Duplicated`](@ref) the gradient is accumulated in-place into `dx`, 
> and a value of `nothing` is placed in the corresponding slot of the returned `Tuple`.

Reviewer: ```suggestion
(for immutable inputs like `x::Float64`) returns them as `derivs`, while the version with `Duplicated` (for mutable inputs like `x::Vector{Float64}`) instead writes the gradient into the `Duplicated` object, and returns `nothing` in the corresponding slot of the returned `derivs`. 
In fact it accumulates the gradient, i.e. if you run it again it will double `dx`.
In general, `autodiff` accepts any mix of `Active` and `Duplicated` function arguments, as well as `Const` and various other `Annotation` types.

the gradient is returned and with [`Duplicated`](@ref) the gradient is accumulated in-place into `dx`, 
and a value of `nothing` is placed in the corresponding slot of the returned `Tuple`.
```

 1.0
 1.0

> julia> autodiff(ForwardWithPrimal, rosenbrock_inp, Duplicated(x, dx))

Reviewer: ```suggestion
julia> autodiff(ForwardWithPrimal, rosenbrock, Duplicated(x, dx))
```
